src/Stratis.Bitcoin.Features.BlockStore/BlockStoreBehavior.cs

```diff
@@ -393,91 +387,99 @@ private Task ProcessSendCmpctPayload(NetworkPeer peer, SendCmpctPayload sendCmpc
 
-                    chainBehavior.SetPendingTip(bestIndex);
-                    await peer.SendMessageAsync(new HeadersPayload(headers.ToArray())).ConfigureAwait(false);
-                    this.logger.LogTrace("(-)[SEND_HEADERS_PAYLOAD]");
-                    return;
-                }
-                else
-                {
-                    revertToInv = true;
+                        chainBehavior.SetPendingTip(bestIndex);
+                        await peer.SendMessageAsync(new HeadersPayload(headers.ToArray())).ConfigureAwait(false);
+                        this.logger.LogTrace("(-)[SEND_HEADERS_PAYLOAD]");
+                        return;
+                    }
+                    else
+                    {
+                        revertToInv = true;
+                    }
                 }
-            }
 
-            if (revertToInv)
-            {
-                // If falling back to using an inv, just try to inv the tip.
-                // The last entry in 'blocksToAnnounce' was our tip at some point in the past.
-
-                if (blocksToAnnounce.Any())
+                if (revertToInv)
                 {
-                    ChainedBlock chainedBlock = blocksToAnnounce.Last();
-                    if (chainedBlock != null)
+                    // If falling back to using an inv, just try to inv the tip.
+                    // The last entry in 'blocksToAnnounce' was our tip at some point in the past.
+
+                    if (blocksToAnnounce.Any())
                     {
-                        if ((chainBehavior.PendingTip == null) || (chainBehavior.PendingTip.GetAncestor(chainedBlock.Height) == null))
+                        ChainedBlock chainedBlock = blocksToAnnounce.Last();
+                        if (chainedBlock != null)
                         {
-                            inventoryBlockToSend.Add(chainedBlock.HashBlock);
-                            this.logger.LogDebug("Sending inventory hash '{0}' to peer '{1}'.", chainedBlock.HashBlock, peer.RemoteSocketEndpoint);
+                            if ((chainBehavior.PendingTip == null) || (chainBehavior.PendingTip.GetAncestor(chainedBlock.Height) == null))
+                            {
+                                inventoryBlockToSend.Add(chainedBlock.HashBlock);
+                                this.logger.LogDebug("Sending inventory hash '{0}' to peer '{1}'.", chainedBlock.HashBlock, peer.RemoteSocketEndpoint);
+                            }
                         }
                     }
                 }
-            }
 
-            if (inventoryBlockToSend.Any())
+                if (inventoryBlockToSend.Any())
+                {
+                    await this.SendAsBlockInventoryAsync(peer, inventoryBlockToSend).ConfigureAwait(false);
+                    this.logger.LogTrace("(-)[SEND_INVENTORY]");
+                    return;
+                }
+            }
+            catch (OperationCanceledException)
```

Comment:
I would say `OperationCanceledException` can be caught on the caller not at each behaviour.